src/irmin/tree.ml

```diff
@@ -1082,113 +1076,120 @@ module Make (P : Private.S) = struct
                 | `Node n -> Lwt.return ((k, `Node n) :: acc))
               [] (List.rev l))
 
-  let may_remove t k =
-    Node.findv t k >>= function
-    | None -> Lwt.return_none
-    | Some _ -> Node.remove t k >>= Lwt.return_some
-
   let empty = `Node (Node.of_map StepMap.empty)
 
   let empty_node = function
     | `Node n -> Node.empty n
     | `Contents _ -> Node.of_map StepMap.empty
 
-  let remove t k =
-    Log.debug (fun l -> l "Tree.remove %a" pp_path k);
-    let empty = empty_node t in
-    match Path.rdecons k with
-    | None ->
-        let* is_empty = is_empty t in
-        if is_empty then Lwt.return t else Lwt.return (`Node empty)
-    | Some (path, file) -> (
-        let rec aux view path k =
-          let some n = k (Some n) in
-          match Path.decons path with
-          | None -> may_remove view file >>= k
-          | Some (h, p) -> (
-              Node.findv view h >>= function
-              | None | Some (`Contents _) -> k None
-              | Some (`Node child) ->
-                  (aux [@tailcall]) child p (function
-                    | None -> k None
-                    | Some child' -> (
-                        (* remove empty dirs *)
-                        Node.is_empty child' >|= get_ok
-                        >>= function
-                        | true -> may_remove view h >>= k
-                        | false -> Node.add view h (`Node child') >>= some)))
-        in
-        let n = match t with `Node n -> n | _ -> empty in
-        (aux [@tailcall]) n path @@ function
-        | None -> Lwt.return t
-        | Some n -> Lwt.return (`Node n))
+  (** During recursive updates, we keep track of whether or not we've made a
+      modification in order to avoid unnecessary allocations of identical tree
+      objects. *)
+  type 'a updated = Changed of 'a | Unchanged
 
-  let add_tree t k v =
-    Log.debug (fun l -> l "Tree.add_tree %a" pp_path k);
-    let empty = empty_node t in
-    match Path.rdecons k with
-    | None -> Lwt.return v
+  let maybe_equal (x : t) (y : t) =
+    if x == y then True
+    else
+      match (x, y) with
+      | `Node x, `Node y -> Node.maybe_equal x y
+      | `Contents x, `Contents y -> if contents_equal x y then True else False
+      | _ -> False
+
+  let update_tree root_tree path f =
+    let empty_node = empty_node root_tree in
+    match Path.rdecons path with
+    | None -> (
+        let* empty_tree =
+          is_empty root_tree >|= function
+          | true -> root_tree
+          | false -> `Node empty_node
+        in
+        match f (Some root_tree) with
+        (* Here we consider "deleting" a root contents value or node to consist
+           of changing it to an empty node. Note that this introduces
+           sensitivity to ordering of subtree operations: updating in a subtree
+           and adding the subtree are not necessarily commutative. *)
+        | None -> Lwt.return empty_tree
+        | Some (`Node _ as t) -> Lwt.return t
```

Comment:
Does it make sense to check for equality between old and new here too?

Reply:
It does, thanks!